### PR TITLE
[primTorch] Implement NLL loss reference

### PR DESCRIPTION
**Pull request by rdspring1**
Add Reference:
- nll_loss

Depends on:
- expand https://github.com/pytorch/pytorch/pull/79820
- advance indexing

### CODE REVIEW

**Comment by IvanYashchuk** [comment]
https://github.com/pytorch/pytorch/pull/79820 was merged. Is advanced indexing still a blocker? What exactly doesn't work?

**Comment by rdspring1** [comment]
https://github.com/pytorch/pytorch/pull/81128/files#diff-93c7b95139f636278cc494028e322a2c3c3c9ba1e83b2adb35d54ccabed5b47aR431
Wouldn't ^^^^ this require advance indexing in primTorch?

File: torch/_refs/nn/functional/__init__.py [hunk 1]
```diff
@@ -419,6 +420,122 @@ def hinge_embedding_loss(
...
 
+def _nll_loss_nd(
+    input: TensorLikeType,
+    target: TensorLikeType,
+    weight: Optional[TensorLikeType],
+    reduction: str,
+    ignore_index: int,
+) -> TensorLikeType:
+    if input.ndim == 3 or input.ndim > 4:
+        msg = "Expected input dimension to be either [1, 2, 4] but recieved {}."
```
Comment (mruberry) [inline torch/_refs/nn/functional/__init__.py:431]: use `utils.check` instead of directly raising

File: torch/_refs/nn/functional/__init__.py [hunk 2]
```diff
@@ -419,6 +420,122 @@ def hinge_embedding_loss(
...
+    if input.ndim != 1 and input.shape[0] != target.shape[0]:
+        msg = "Expected input batch size ({}) to match target batch size ({})."
+        raise ValueError(msg.format(input.shape[0], target.shape[0]))
+
+    _check_reduction_value(reduction)
+
+    flat_target = torch.reshape(target, [-1])
+    ignore_classes_mask = torch.eq(flat_target, ignore_index)
+    ignore_class_weight = torch.scalar_tensor(0, dtype=input.dtype, device=input.device)
+    default_class_weight = torch.scalar_tensor(
```
Comment (mruberry) [inline torch/_refs/nn/functional/__init__.py:443]: This is only used in the `if weight is None` branch below, so it should only be constructed in that branch, too

File: torch/_refs/nn/functional/__init__.py [hunk 3]
```diff
@@ -419,6 +420,122 @@ def hinge_embedding_loss(
...
+    input: TensorLikeType,
+    target: TensorLikeType,
+    weight: Optional[TensorLikeType] = None,
+    size_average: Optional[bool] = None,
+    ignore_index: int = -100,
+    reduce: Optional[bool] = None,
+    reduction: str = "mean",
+) -> TensorLikeType:
+    if size_average is not None or reduce is not None:
+        # TODO: raise exception instead of converting value
```
Comment (mruberry) [inline torch/_refs/nn/functional/__init__.py:505]: What's the current state of the actual `nll_loss` operation?

File: torch/_refs/nn/functional/__init__.py [hunk 4]
```diff
@@ -419,6 +420,122 @@ def hinge_embedding_loss(
...
+        # msg = "size_average and reduce args are deprecated, please use reduction argument."
+        reduction = _get_string_reduction_arg(size_average=size_average, reduce=reduce)
+
+    if input.ndim == 3 or input.ndim > 4:
+        # input ndim is == 3 or > 4
```
Comment (mruberry) [inline torch/_refs/nn/functional/__init__.py:510]: This comment just repeats the previous line

File: torch/_refs/nn/functional/__init__.py [hunk 5]
```diff
@@ -419,6 +420,122 @@ def hinge_embedding_loss(
...
+            target = torch.reshape(target, [batch_size, 1, -1])
+        else:
+            target = torch.reshape(target, [batch_size, 0, 0])
+
+        if reduction == "none":
+            return _nll_loss_nd(input, target, weight, reduction, ignore_index)
+        else:
+            result = _nll_loss_nd(input, target, weight, reduction, ignore_index)
+            return torch.reshape(result, out_size)
+    else:
```
Comment (mruberry) [inline torch/_refs/nn/functional/__init__.py:535]: Add a comment describing what this else branch is for.

From a code organization and readability standpoing these branches seem a little odd. Maybe we can explain them better?

In particular -- can input be zero or one dimension? If so, how do we interpret that? The documentation for suggests that input should have at least two dimensions. And why are inputs with three or four dimensions special? 

Finally, prefer putting shorter branches which short-circuit first. That typically lets code have fewer indentation levels:

```
# shortcircuits if foo because...
if foo:
  return x

# implicit else branch doesn't have to be indented
...
```

File: torch/_refs/nn/functional/__init__.py [hunk 6]
```diff
@@ -419,6 +420,122 @@ def hinge_embedding_loss(
...
+        if target.numel() > 0:
+            target = torch.reshape(target, [batch_size, 1, -1])
+        else:
+            target = torch.reshape(target, [batch_size, 0, 0])
+
+        if reduction == "none":
+            return _nll_loss_nd(input, target, weight, reduction, ignore_index)
+        else:
+            result = _nll_loss_nd(input, target, weight, reduction, ignore_index)
+            return torch.reshape(result, out_size)
```
Comment (mruberry) [inline torch/_refs/nn/functional/__init__.py:534]: This will need a comment

File: torch/_refs/nn/functional/__init__.py [hunk 7]
```diff
@@ -419,6 +420,122 @@ def hinge_embedding_loss(
...
+        # msg = "size_average and reduce args are deprecated, please use reduction argument."
+        reduction = _get_string_reduction_arg(size_average=size_average, reduce=reduce)
+
+    if input.ndim == 3 or input.ndim > 4:
+        # input ndim is == 3 or > 4
+        batch_size = input.shape[0]
+        num_classes = input.shape[1]
+        out_size = [batch_size] + list(input.shape[2:])
+
+        if target.shape[1:] != input.shape[2:]:
```
Comment (mruberry) [inline torch/_refs/nn/functional/__init__.py:515]: `utils.check`

File: torch/_refs/nn/functional/__init__.py [hunk 8]
```diff
@@ -419,6 +420,122 @@ def hinge_embedding_loss(
...
+            msg = "Expected target size {} but got {}"
```
Comment (mruberry) [inline torch/_refs/nn/functional/__init__.py:516]: "size" -> "shape"

File: torch/_refs/nn/functional/__init__.py [hunk 9]
```diff
@@ -419,6 +420,122 @@ def hinge_embedding_loss(
...
+    # TODO: This check does not work with FakeTensor inputs
+    """
+    num_classes = input.shape[1] if input.ndim > 1 else input.shape[0]
+    valid_classes_mask = torch.logical_and((flat_target >= 0), (flat_target < num_classes))
+    if not torch.all(torch.logical_or(ignore_classes_mask, valid_classes_mask)):
+        print(target, num_classes, ignore_index)
+        raise ValueError("Target class is out-of-bounds and not ignore index")
+    """
+
+    if weight is None:
```
Comment (mruberry) [inline torch/_refs/nn/functional/__init__.py:456]: This will need a comment, particularly to explain the expansion.

Instead of branching here it might be better to prepare the arguments before calling `where`

File: torch/_refs/nn/functional/__init__.py [hunk 10]
```diff
@@ -419,6 +420,122 @@ def hinge_embedding_loss(
...
+            default_class_weight,
+        )
+    else:
+        current_weight = torch.where(
+            ignore_classes_mask,
+            ignore_class_weight.expand_as(flat_target),
+            weight[flat_target],
+        )
+
+    if input.ndim == 1:
```
Comment (mruberry) [inline torch/_refs/nn/functional/__init__.py:469]: Add comments describing these cases

File: torch/_refs/nn/functional/__init__.py [hunk 11]
```diff
@@ -419,6 +420,122 @@ def hinge_embedding_loss(
...
+        hdx = (torch.arange(numel) - (bdx * extent)) // width
+        wdx = torch.arange(numel) % width
+        loss = -input[bdx, flat_target, hdx, wdx] * current_weight
+    loss = torch.reshape(loss, target.shape)
+
+    if reduction == "none":
+        return loss
+    elif reduction == "sum":
+        return torch.sum(loss)
+    else:
```
Comment (mruberry) [inline torch/_refs/nn/functional/__init__.py:490]: Comment for this case

File: torch/_refs/nn/functional/__init__.py [hunk 12]
```diff
@@ -419,6 +420,122 @@ def hinge_embedding_loss(
...
+
+    flat_target = torch.reshape(target, [-1])
+    ignore_classes_mask = torch.eq(flat_target, ignore_index)
+    ignore_class_weight = torch.scalar_tensor(0, dtype=input.dtype, device=input.device)
+    default_class_weight = torch.scalar_tensor(
+        1, dtype=input.dtype, device=input.device
+    )
+
+    # TODO: This check does not work with FakeTensor inputs
+    """
```
Comment (mruberry) [inline torch/_refs/nn/functional/__init__.py:448]: These data-dependent checks are a pain. 

Would you file an issue so we can discuss this as a group?

File: torch/_refs/nn/functional/__init__.py (hunk 3 above)
Comment (rdspring1) [reply torch/_refs/nn/functional/__init__.py:509]: Currently it returns `UserWarning: size_average and reduce args will be deprecated, please use reduction='mean' instead.` when `reduce=True`

File: torch/_refs/nn/functional/__init__.py [hunk 13]
```diff
@@ -419,6 +420,122 @@ def hinge_embedding_loss(
...
+        # input ndim is == 3 or > 4
+        batch_size = input.shape[0]
+        num_classes = input.shape[1]
+        out_size = [batch_size] + list(input.shape[2:])
+
+        if target.shape[1:] != input.shape[2:]:
+            msg = "Expected target size {} but got {}"
+            raise ValueError(msg.format(out_size, target.shape))
+
+        # support empty batches, see #15870
```
[the remark this answers is not in the document]
Comment (rdspring1) [reply torch/_refs/nn/functional/__init__.py]: I added a short-circuit for empty tensors.
```
    # The expected behavior when the target and input have zero elements:
    #   reduction = 'none' --- tensor([])
    #   reduction = 'sum'  --- tensor(0.)
    #   reduction = 'mean' --- tensor(nan)
    # Mean reduction on empty tensors produces NaN. See the discussion in
    # https://github.com/pytorch/pytorch/pull/64572#issuecomment-926504162
    if input.numel() == 0 and target.numel() == 0:
        if reduction == "none":
            return torch.zeros_like(target)
        elif reduction == "sum":
            return torch.empty_like(target)
        else:
            return torch.full_like(target, float("nan"))
```

File: torch/_refs/nn/functional/__init__.py (hunk 5 above)
Comment (rdspring1) [reply torch/_refs/nn/functional/__init__.py]: I refactored _nll_loss_nd to handle 1-3 dimensions. If there are more than 3 dimensions, the k-dimension is flattened to create a 3D tensor. The Aten implementation used a 4D case for image inputs.
```
    # The _nll_loss_nd helper function handles the most common cases.
    # ndim == 1 (Single Example)
    #   => Batch Size: 1, Input: (C), Target: ()
    # ndim == 2 (k = 1)
    #   => Batch Size: N, Input: (N, C), Target: (N)
    # ndim == 3 (k > 1)
    #   => Batch Size: N, Input: (N, C, K), Target: (N, K)
    # ndim > 3
    #   => reshape the input and target to the 3-D case
```

File: torch/_refs/nn/functional/__init__.py (hunk 3 above)
Comment (mruberry) [reply torch/_refs/nn/functional/__init__.py:516]: Let's include that in this comment so a reader understands this behavior is for consistency with the current eager operation

File: torch/_refs/nn/functional/__init__.py (hunk 5 above)
Comment (mruberry) [reply torch/_refs/nn/functional/__init__.py]: Is the 4D case interesting to model here?

File: torch/_refs/nn/functional/__init__.py [hunk 14]
```diff
@@ -432,6 +434,152 @@ def hinge_embedding_loss(
...
+        (flat_target >= 0), (flat_target < num_classes)
+    )
+    class_check = torch.all(torch.logical_or(ignore_classes_mask, valid_classes_mask))
+
+    # TODO: This check does not work with FakeTensor inputs; See Issue #85834
+    # Explicit cast for class_check to bool; See Issue #78071
+    utils.check(
+        isinstance(target, FakeTensor) or bool(class_check.item()),
+        lambda: "A target class is out-of-bounds and not the ignore index.",
+    )
```
Comment (IvanYashchuk) [inline torch/_refs/nn/functional/__init__.py:470]: Comment this out for now until we have a debug mode for data-dependent checks.

File: torch/_refs/nn/functional/__init__.py [hunk 15]
```diff
@@ -432,6 +434,152 @@ def hinge_embedding_loss(
...
+    if input.ndim == 1:
+        # implicit batch size = 1
+        # input (1 batch size, C classes)
+        loss = -input[target] * current_weight
+    elif input.ndim == 2:
+        # input (N batch size, C classes)
+        batch_size = input.shape[0]
+        loss = -input[torch.arange(batch_size), target] * current_weight
+    else:
+        # 3D case (N batch size, C classe, K dimensions)
```
Comment (mruberry) [inline torch/_refs/nn/functional/__init__.py:493]: ND case

**Review by mruberry — APPROVED** [review]
Cool! Let's just update the data-dependent check per @IvanYashchuk's comment

File: torch/_refs/nn/functional/__init__.py [hunk 16]
```diff
@@ -432,6 +434,152 @@ def hinge_embedding_loss(
...
+        return loss
+    elif reduction == "sum":
+        return torch.sum(loss)
+    else:
+        # calculate weighted mean of the loss function
+        return torch.sum(loss) / torch.sum(current_weight)
+
+
+@register_decomposition(torch.ops.aten.nll_loss)
+def nll_loss(
```
Comment (mruberry) [inline torch/_refs/nn/functional/__init__.py:514]: try wrapping with type promotion decorator
